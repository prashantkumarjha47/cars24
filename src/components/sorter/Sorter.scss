.sorter {
  width: 100%;
  .row {
    display: flex;
    justify-content: flex-end;
    .search-row {
      @media screen and (max-width: 991px) {
        margin-bottom: 10px;
      }
    }
    .filter-btn {
      display: none;
      @media screen and (max-width: 991px) {
        display: block;
        margin-left: auto;
      }
    }
  }
}
