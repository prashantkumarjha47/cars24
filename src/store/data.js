const data = {
  carCopy: [
    {
      id: 1,
      title: "Tata Nexon",
      price: 50000,
      category: "Compact SUV",
      image: "...",
      color: "white",
      condition: "New",
      buildYear: "2020",
      make: "TATA",
      model: "XE",
    },
    /*...*/
    {
      id: 2,
      title: "Kia Sonet",
      price: 225000,
      category: "Compact SUV",
      image: "...",
      color: "black",
      condition: "new",
      buildYear: "2020",
      make: "Kia",
      model: "1.2 HTE 5MT",
    },
    {
      id: 3,
      title: "Ford EcoSport",
      price: 300000,
      category: "Compact SUV",
      image: "...",
      color: "blue",
      condition: "new",
      buildYear: "2020",
      make: "Ford",
      model: "1.5L Petrol Ambiente MT",
    },
    {
      id: 4,
      title: "Hyundai Venue",
      price: 1675000,
      category: "Compact SUV",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 5,
      title: "Hyundai Venue",
      price: 875000,
      category: "Compact SUV",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 6,
      title: "Hyundai Venue",
      price: 500000,
      category: "Compact SUV",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 7,
      title: "Hyundai Venue",
      price: 2575000,
      category: "Compact SUV",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 8,
      title: "Hyundai Venue",
      price: 600000,
      category: "Compact SUV",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 9,
      title: "Hyundai Venue",
      price: 150000,
      category: "Compact SUV",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 10,
      title: "Hyundai Venue",
      price: 480000,
      category: "Compact SUV",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 11,
      title: "Hyundai Venue",
      price: 955000,
      category: "Compact SUV",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 12,
      title: "Hyundai Venue",
      price: 740000,
      category: "Compact SUV",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 13,
      title: "Hyundai Venue",
      price: 520000,
      category: "Premium hatchback",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 15,
      title: "Hyundai Venue",
      price: 98000,
      category: "Premium hatchback",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 16,
      title: "Hyundai Venue",
      price: 375000,
      category: "Hatchback",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 17,
      title: "Hyundai Venue",
      price: 347000,
      category: "Hatchback",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 18,
      title: "Hyundai Venue",
      price: 552000,
      category: "Hatchback",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 19,
      title: "Hyundai Venue",
      price: 614000,
      category: "Hatchback",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 20,
      title: "Hyundai Venue",
      price: 1005000,
      category: "Hatchback",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 21,
      title: "Hyundai Venue",
      price: 75000,
      category: "Hatchback",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
  ],
  categories: [
    {
      id: 0,
      title: "",
    },
    {
      id: 1,
      title: "Compact SUV",
    },
    {
      id: 2,
      title: "Premium hatchback",
    },
    {
      id: 3,
      title: "Hatchback",
    },
  ],
  condition: [
    {
      id: 0,
      title: "",
    },
    {
      id: 1,
      type: "used",
    },
    {
      id: 2,
      type: "damaged",
    },
  ],
  makes: [
    {
      id: 0,
      title: "",
    },
    {
      id: 1,
      name: "TATA",
    },
    {
      id: 2,
      name: "Kia",
    },
    {
      id: 3,
      name: "Hyundai",
    },
    {
      id: 4,
      name: "Ford",
    },
    {
      id: 5,
      name: "Maruti",
    },
  ],
};

export default data;
