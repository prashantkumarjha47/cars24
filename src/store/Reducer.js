import data from "./data";

const OFFSET = 4;
const arr = data.carCopy.sort((a, b) => a.price - b.price);
const initialState = {
  ...data,
  n: OFFSET,
  cars: data.carCopy.slice(0, OFFSET),
  minVal: arr[0].price,
  maxValu: arr[arr.length - 1].price,
  cart: [],
};

const reducer = (state, action) => {
  switch (action.type) {
    case "GET_CARS":
      const { category, priceRange, vehicleCondition, make, search } = state;
      let filteredCars = state.carCopy
        .filter((car) => {
          return (
            (!category || car.category === category) &&
            (!priceRange || car.price <= priceRange) &&
            (!vehicleCondition || car.vehicleCondition === vehicleCondition) &&
            (!make || car.make === make) &&
            (!search || car.title.toLowerCase().includes(search.toLowerCase()))
          );
        })
        .slice(0, state.n + OFFSET);

      if (state.sort) {
        filteredCars = filteredCars.sort((a, b) =>
          state.sort === "ASC"
            ? parseFloat(a.price) - parseFloat(b.price)
            : parseFloat(b.price) - parseFloat(a.price)
        );
      }

      return {
        ...state,
        n: state.n + OFFSET,
        cars: filteredCars,
      };

    case "GET_CATEGORIES":
      return {
        ...state,
        categories: [...state.categories],
      };
    case "GET_CONDITION":
      return {
        ...state,
        condition: [...state.condition],
      };
    case "SEARCH_CAR":
      return {
        ...state,
        search: action.payload,
        cars: state.cars.filter((car) =>
          car.title?.toLowerCase().includes(action.payload?.toLowerCase())
        ),
      };
    case "SORT_CAR_ASC":
      return {
        ...state,
        sort: "ASC",
        cars: state.cars.sort(
          (a, b) => parseFloat(a.price) - parseFloat(b.price)
        ),
      };
    case "SORT_CAR_DESC":
      return {
        ...state,
        sort: "DESC",
        cars: state.cars.sort(
          (a, b) => parseFloat(b.price) - parseFloat(a.price)
        ),
      };
    case "APPLY_FILTER":
      return {
        ...state,
        n: OFFSET,
        ...action.payload,
        cars: [],
      };
    case "BOOK_NOW":
      return {
        ...state,
        cart: [...state.cart, action.payload],
      };

    default:
      return state;
  }
};

export { reducer, initialState };
