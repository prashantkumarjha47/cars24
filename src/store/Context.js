import React, { useReducer } from "react";
import { reducer, initialState } from "./Reducer";

const Context = React.createContext();

const ContextProvider = (props) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <Context.Provider value={[state, dispatch]}>
      {props.children}
    </Context.Provider>
  );
};

export { ContextProvider, Context };
