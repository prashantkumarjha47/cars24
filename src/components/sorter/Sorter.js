import React, { useState, useContext } from "react";
import { Container, Row, Col, Input, Button } from "reactstrap";
import { Context } from "../../store/Context";

import "./Sorter.scss";

function Sorter(props) {
  const [searchTxt, setSearchTxt] = useState("");
  const [sortCondition, setSortCondition] = useState("");

  const [, dispatch] = useContext(Context);

  const onSearch = (e) => {
    setSearchTxt(e.target.value);
    dispatch({
      type: "SEARCH_CAR",
      payload: e.target.value,
    });
    dispatch({ type: "GET_CARS" });
  };

  const onSort = (e) => {
    const order = e.target.value;
    setSortCondition(order);
    dispatch({
      type: `SORT_CAR_${order}`,
    });
  };

  return (
    <div className="sorter">
      <Container>
        <Row>
          <Col xs="12" sm="12" md="12" lg="4" className="search-row">
            <Input onChange={onSearch} value={searchTxt} placeholder="Search" />
          </Col>
          <Col xs="7" sm="8" md="8" lg="4">
            <Input
              onChange={onSort}
              type="select"
              placeholder="Filter by price"
              value={sortCondition}
            >
              <option value="ASC">Low to High</option>
              <option value="DESC">High to Low</option>
            </Input>
          </Col>
          <Col className="filter-btn" xs={{ size: "auto" }}>
            <Button onClick={props.onFilterClick} color="info">
              Filter
            </Button>{" "}
          </Col>
        </Row>
      </Container>
    </div>
  );
}

export default React.memo(Sorter);
