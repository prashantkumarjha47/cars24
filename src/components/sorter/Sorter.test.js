import renderer from "react-test-renderer";
import Sorter from "./Sorter";
import { ContextProvider } from "../../store/Context";

test("Filter component loaded", () => {
  const component = renderer.create(
    <ContextProvider>
      <Sorter onFilterClick={() => {}} />
    </ContextProvider>
  );
  expect(component.toJSON()).toMatchSnapshot();
});
