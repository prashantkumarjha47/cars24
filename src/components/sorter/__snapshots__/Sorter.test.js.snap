// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter component loaded 1`] = `
<div
  className="sorter"
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="search-row col-12 col-sm-12 col-md-12 col-lg-4"
      >
        <input
          className="form-control"
          onChange={[Function]}
          placeholder="Search"
          type="text"
          value=""
        />
      </div>
      <div
        className="col-7 col-sm-8 col-md-8 col-lg-4"
      >
        <select
          className="form-control"
          onChange={[Function]}
          placeholder="Filter by price"
          value=""
        >
          <option
            value="ASC"
          >
            Low to High
          </option>
          <option
            value="DESC"
          >
            High to Low
          </option>
        </select>
      </div>
      <div
        className="filter-btn col-auto"
      >
        <button
          aria-label={null}
          className="btn btn-info"
          onClick={[Function]}
          type="button"
        >
          Filter
        </button>
         
      </div>
    </div>
  </div>
</div>
`;
